File: StriverA2Z/Java/Arrays/Hard/MajorityNBy3.java
import java.util.ArrayList;

public class MajorityNBy3 {
    public static void main(String[] args) {
        int[] arr = {2, 2, 1, 3, 1, 1, 3, 1, 1};
        majorityNBy3(arr);
    }

    static void majorityNBy3(int[] arr) {
        int cnt1 = 0, cnt2 = 0, el1 = Integer.MIN_VALUE, el2 = Integer.MIN_VALUE;
        for (int i : arr) {
            if (cnt1 == 0 && el2 != i) {
                cnt1 = 1;
                el1 = i;
            } else if (cnt2 == 0 && el1 != i) {
                el2 = i;
                cnt2 = 1;
            } else if (el1 == i) {
                cnt1++;
            } else if (el2 == i) {
                cnt2++;
            } else {
                cnt1--;
                cnt2--;
            }
        }

        cnt1 = 0;
        cnt2 = 0;

        for (int i : arr) {
            if (el1 == i) cnt1++;
            if (el2 == i) cnt2++;
        }

        int min = (arr.length / 3) + 1;
        ArrayList<Integer> ls = new ArrayList<>();
        if (cnt1 >= min) ls.add(el1);
        if (cnt2 >= min) ls.add(el2);

        for (Integer i : ls) {
            System.out.println(i);
        }
    }
}

File: StriverA2Z/JavaScript&TypeScript/src/Arrays/Hard/majorityNBy3.ts
function majorityNBy3(arr: number[]) {
	let cnt1 = 0, cnt2 = 0, el1, el2;
	for (let i of arr) {
		if (cnt1 === 0 && el2 !== i) {
			cnt1 = 1;
			el1 = i;
		} else if (cnt2 === 0 && el1 !== i) {
			cnt2 = 1;
			el2 = i;
		} else if (el1 === i) {
			cnt1++;
		} else if (el2 === i) {
			cnt2++;
		} else {
			cnt1--;
			cnt2--;
		}
	}

	cnt1 = 0;
	cnt2 = 0;

	const ls: number[] = [];
	for (let i of arr) {
		if (i === el1) cnt1++;
		if (i === el2) cnt2++;
	}

	const min = Math.floor(arr.length / 3) + 1;
	if (cnt1 >= min) ls.push(el1);
	if (cnt2 >= min) ls.push(el2);
	console.log(ls);
}

majorityNBy3([2, 2, 1, 3, 1, 1, 3, 1, 1])